### ISSUE TITLE
Parameter extensions / Parameter types for resource types.

### ISSUE THREAD

**Issue by benbridts** [aws-cloudformation/cloudformation-coverage-roadmap#952]
### Name of the resource

Other

### Description

Currently CloudFormation supports a limited list of [AWS-specific parameter types]( https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/parameters-section-structure.html#aws-specific-parameter-types). It would be a very useful to have parameter types for:
- every resource type that implements a List handler
- private parameter types

It could even go a step further and resolv attributes of those resources

Examples:

```
# Returns the primary identifier
Type: AWS::CloudFormation::Stack::$id
# Selection options for multiple primary identifiers
Type: List<AWS::CloudFormation::Stack::$id>
# Using a private resource type
Type: Type: MyCompany::Service::Product::$id
# Returns the !GetAtt StackSetId of the selected resource (selection still via primary identifier)
# Note: this might be confusing, but querying all Attributes seems unfeasable
Type: AWS::CloudFormation::StackSet::StackSetId
Type: List<AWS::CloudFormation::StackSet::StackSetId>
```

Of course alternative syntax is also possible.
Eg.
```
# primary identifiers
Type: CloudControl<AWS::CloudFormation::Stack>
Type: CloudControlList<AWS::CloudFormation::Stack>
# GetAtt
Type: CloudControlAttribute<AWS::CloudFormation::StackSet, StackSetId>
Type: CloudControlAttributeList<AWS::CloudFormation::StackSet, StackSetId>

### Other Details

This is probably extra powerful when used in Service Catalog

**Comment by benkehoe** [comment]
I don't think this should require types that implement list handlers, nor should it need a specification of a property within a resource type. The parameter type should just be *the resource type itself* and the input value required in the stack operation is the primary identifier as defined in the schema, and then the input name is referenceable like a normal resource. The list handler, if it's defined, is used by the AWS console to provide a dropdown of primary identifiers.

The template would look like:
```yaml
Parameters:
  Role:
    Type: AWS::IAM::Role
Resources:
  Function:
    Type: AWS::Lambda::Function
    Properties:
      Role: !GetAtt Role.Arn
      # etc
```
and then `CreateStack` requires a role ARN as a parameter value.

Same with lists of types, e.g. `List<AWS::IAM::Role>`; a syntax for referencing resources within the list could be common with the `Fn::Map` proposal in #941.

**Comment by lejiati** [comment]
@benbridts  Thank you very much for your feedback! Since this repository is focused on resource coverage, I'm transferring this issue over to a new GitHub repository dedicated to CloudFormation template language issues.
